docs/book/src/blockchain-development/external_code.md

```diff
@@ -0,0 +1,74 @@
+# External Code Execution
+
+The `std-lib` includes a function called `run_external` that allows Sway contracts to execute arbitrary external Sway code.
+
+This functionality enables features like upgradeable contracts and
+proxies.
+
+## Upgradeable Contracts
+
+Upgradeable contracts are designed to allow the logic of a smart contract to be updated after deployment.
+
+Consider this example proxy contract:
+
+```sway
+{{#include ../../../../examples/upgradeable_proxy/proxy/src/main.sw:proxy}}
+```
+
+The contract has two functions:
+
+- `set_target_contract` updates the `target_contract` variable in storage with the `ContractId` of an external contract.
+- `double_input` reads the `target_contract` from storage and uses it to run external code. If the `target_contract` has a function with the same name (`double_input`), the code in the external `double_input` function will run.
+In this case, the function will return a `u64`.
+
+Below is what an implementation contract could look like for this:
+
+```sway
+{{#include ../../../../examples/upgradeable_proxy/implementation/src/main.sw:target}}
+```
+
+This contract has one function called `double_input`, which calculates the input value times two, updates the `value` variable in storage, and returns the new value.
+
+## How does this differ from calling a contract?
+
+There are a couple of major differences between calling a contract directly and using the `run_external` method.
+
+First, to use `run_external`, the ABI of the external contract is not required. The proxy contract has no knowledge of the external contract except for its `ContractId`.
+
+### Upgradeable Contract Storage
+
+Second, the storage context of the proxy contract is retained for the loaded code.
+This means that in the examples above, if someone calls the `double_input` function on the proxy contract, the `value` variable will be updated in the proxy contract's storage.
```

Comment:
The example is wrong. The `value` variable in the proxy storage will not get updated, and it's because it's placed at a completely different location in the storage because of the namespace attribute.

From our conversation I think that you understand which storage gets updated, but you seem to be under the impression that a variable needs to exist in the proxy's storage for the target to modify it. This is not the case.

Were the `value` variable not declared in the proxy's storage (and in effect it isn't because the one you have here is moved by the namespace at a different location) the target code would still work and update slots in the proxy's storage *as if* it were laid out in the way that is described by the target's storage specification.